src/helpers/transform-avatar.js

```diff
@@ -0,0 +1,72 @@
+/**
+@file transform avatar
+@summary Helpers for flipping and rotating the avatar sprite
+*/
+
+import bitsy from "bitsy";
+import {
+	getSpriteData,
+	setSpriteData
+} from "./edit image at runtime";
+
+// copied from https://stackoverflow.com/a/46805290
+function transpose(matrix) {
+  const rows = matrix.length, cols = matrix[0].length;
+  const grid = [];
+  for (let j = 0; j < cols; j++) {
+    grid[j] = Array(rows);
+  }
+  for (let i = 0; i < rows; i++) {
+    for (let j = 0; j < cols; j++) {
+      grid[j][i] = matrix[i][j];
+    }
+  }
+  return grid;
+}
+
+function transformSpriteData(spriteData, flipVertically, flipHorizontally, rotate) {
+	var x, y, x2, y2, col, tmp;
+	var s = spriteData.slice();
+	if (flipVertically) {
+		for (y = 0; y < s.length / 2; ++y) {
+			y2 = s.length - y - 1;
+			tmp = s[y];
+			s[y] = s[y2];
+			s[y2] = tmp;
+		}
+	}
+	if (flipHorizontally) {
+		for (y = 0; y < s.length; ++y) {
+			col = s[y] = s[y].slice();
+			for (x = 0; x < col.length / 2; ++x) {
+				x2 = col.length - x - 1;
+				tmp = col[x];
+				col[x] = col[x2];
+				col[x2] = tmp;
+			}
+		}
+	}
+	if (rotate) {
+		s = transpose(s);
+	}
+	return s;
+}
+
+export function transformAvatar(originalAnimation, vflip, hflip, rotate) {
```

Comment:
i think maybe you misinterpreted what i meant when i said the shared code from directional avatar should be pulled out into a helper; the `transpose` and `transformSpriteData` functions make sense as-is, but `transformAvatar` doesn't look like it'll work as intended: `originalAnimation` here will never actually be cached b/c it's being passed in as an argument and then assigned to in the case where it doesn't exist. the caller will never see that defaulted value, so it'll end up doing this every time as written. my expectation was that this bit (caching original + applying transforms) would still live in the hacks, and just the helpers above would be abstracted out

Reply:
if you'd like to keep this whole thing abstracted out, it's possible to do that, but the cache would need to live in the scope of this file (which would likely introduce a lot of complexity if it supported any target, but since it is just for the avatar as written it wouldn't be too bad)